### ISSUE TITLE
Provide secure way of identifying customer

### ISSUE THREAD

**Issue by robserva** [papercups-io/papercups#803]
**Problem**
As a user, I can change the email address associated with my Papercups chat client-side. This would lead whoever is chatting with me to believe I'm someone else and if they aren't doing other forms of verification they might reveal some information about me.

**Solution**
I'd like to be able to do one of the following:
- start a papercups "session" server side, where I identify the user's email address, and other customer information. Then receive a session token (e.g. a UUID) and provide that to the client. This way, the session token can not be guessed by another user.
- or Papercups provides me with an HMAC key (or I set one myself in settings) and then I sign the fields, like the customer's email, and include the HMAC along with that information when initializing the Papercup settings client-side. This way Papercups can verify that the information provided by the client was signed off by the server without a round trip talking to the server.

**Comment by cheeseblubber** [comment]
For reference https://developers.intercom.com/installing-intercom/docs/enable-identity-verification-on-your-web-product

**Comment by reichert621** [comment]
thanks for the write-up @robserva! been meaning to do something like this :+1:

**Comment by lucasavila00** [comment]
That is a feature that would be really useful.

On top of that, it would be really interesting if an identified user could be reliably reached.

So if the user has sent a message or not I can send him one and it will send the message to their browsers.

**Comment by robserva** [comment]
> What signing algorithm to use? I was considering JWT but it seems overkill, as we don't need to use any public key features. An HMAC signature of the parameters encoded as query strings should be enough. Which algorithm to use, then: SHA1 or SHA256?

I agree that JWT isn't needed. HMAC SHA256 should do the trick, and I think that's the general best practice.

In theory HMAC SHA1 should still be secure (https://crypto.stackexchange.com/questions/26510/why-is-hmac-sha1-still-considered-secure) but I think SHA256 is still standard.

**Comment by lucasavila00** [comment]
I imagine a possible implementation would be to validate the metadata/customer params with the signature after calling `Customers.sanitize_metadata()` [here](https://github.com/papercups-io/papercups/blob/master/lib/chat_api_web/controllers/customer_controller.ex#L57)  and [here](https://github.com/papercups-io/papercups/blob/master/lib/chat_api_web/controllers/customer_controller.ex#L144) 

I haven't figured out how to access information specific to the `account_id`, though. I would need to know a boolean which defines the need to check for signatures and the private key needed to check the HMAC.

On the client side it should just pass the string around until it reaches these 2 endpoints.

**Comment by reichert621** [comment]
(bumping this up, see https://app.intercom.com/a/apps/x4uvwkvj/settings/identity-verification/web)
